prow/config/inrepoconfig_test.go

```diff
@@ -631,3 +640,98 @@ func testDefaultProwYAMLGetter_RejectsNonGitHubRepo(clients localgit.Clients, t
 		t.Errorf("Error %v does not have expected message %s", err, expectedErrMsg)
 	}
 }
+
+type testClientFactory struct {
+	git.ClientFactory // This will be nil during testing, we override the functions that are used.
+	clientsCreated    int
+}
+
+func (cf *testClientFactory) ClientFor(org, repo string) (git.RepoClient, error) {
+	cf.clientsCreated++
+	// Returning this RepoClient ensures that only Fetch() is called and that Close() is not.
+	return &fetchOnlyNoCleanRepoClient{}, nil
+}
+
+type fetchOnlyNoCleanRepoClient struct {
+	git.RepoClient // This will be nil during testing, we override the functions that are allowed to be used.
+}
+
+func (rc *fetchOnlyNoCleanRepoClient) Fetch() error {
+	return nil
+}
+
+// TestInRepoConfigGitCacheConcurrency validates the following properties of InRepoConfigGitCache
+// - RepoClients are protected from concurrent use.
+// - Use of a RepoClient for one repo does not prevent concurrent use of a RepoClient for another repo.
+// - RepoClients are reused, only 1 client is created per repo.
+func TestInRepoConfigGitCacheConcurrency(t *testing.T) {
+	t.Parallel()
+
+	cf := &testClientFactory{}
+	cache := NewInRepoConfigGitCache(cf)
+	org, repo1, repo2 := "org", "repo1", "repo2"
+	// block channels are populated from the main thread, signal channels are read from the main thread.
+	block1, signal1, signal2, signal3 := make(chan bool), make(chan bool), make(chan bool), make(chan bool)
+
+	// Thread 1: gets a client for repo1, signals on signal1, then blocks on block1 before Clean()ing the repo1 client.
+	thread1 := func() {
+		client, err := cache.ClientFor(org, repo1)
+		if err != nil {
+			t.Errorf("Unexpected error getting repo client for thread 1: %v.", err)
+			return
+		}
+		defer client.Clean()
+		signal1 <- true
+		<-block1
+	}
+
+	// Thread 2: gets a client for repo1, signals success on signal2, then Clean()s the repo1 client.
+	thread2 := func() {
+		client, err := cache.ClientFor(org, repo1)
+		if err != nil {
+			t.Errorf("Unexpected error getting repo client for thread 2: %v.", err)
+			return
+		}
+		defer client.Clean()
+		signal2 <- true
+	}
+
+	// Thread 3: gets a client for repo2, signals success on signal3, then Clean()s the repo2 client.
+	thread3 := func() {
+		client, err := cache.ClientFor(org, repo2)
+		if err != nil {
+			t.Errorf("Unexpected error getting repo client for thread 3: %v.", err)
+			return
+		}
+		defer client.Clean()
+		signal3 <- true
+	}
+
+	// First start thread1 and wait for it to get a repo1 client
+	go thread1()
+	<-signal1
+	// Now start the other two threads
+	go thread2()
+	go thread3()
+	// Ensure thread3 is not blocked by thread1 (repo2 client isn't blocked by concurrent repo1 client usage)
+	<-signal3
+	if cf.clientsCreated != 2 {
+		t.Errorf("Expected 2 clients to be created at this point, but got %d.", cf.clientsCreated)
+	}
+	// Try to confirm that thread2 is blocked by thread1 (disallow concurrent use of repo1 client)
```

Comment:
We could avoid this by leveraging the race detector in what we purposely introduce a datarace in thread 1 and thread 2 (i.E. make them write to the same variable or sth like that).

This should make the racedector go off if the locking isn't working correctly and might simplify the testcase overall, as we only need to start the three threads, wait for them, verify two clients got created and the racedector didn't go off.

Reply:
Sure, thats a good idea. I figured this kind of explicitly synchronized test would be more reliable/deterministic than depending on the race detector though. Do you know how reliable the race detector is?
Essentially I'm wondering if you think we should add what you are suggesting as a new test and keep the existing test, or replace the existing tests.

Reply:
The racedetector fails whenever a data race is possible and doesn't require one to actually happen, so it is deterministic (will always fail/succeed for the same code). Interestlingly the golang docs do not explicitly say anything about that though:  https://golang.org/doc/articles/race_detector

It works by checking if there are unsynchronized variable accesses where at least one is a write. I couldn't quickly find a blogpost about this but I saw a talk some years ago. Something like https://www.youtube.com/watch?v=5erqWdlhQLA  likely has the details if you are interested.

Reply:
Neat! Well I've updated the test to the simplified form you suggested, PTAL.